clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -351,7 +408,69 @@ private List<NetworkClientDelegate.UnsentRequest> sendListOffsetsRequestsAndRese
                     }
                 }
             });
-        }
+        });
+        return unsentRequests;
+    }
+
+    /**
+     * For each partition which needs validation, make an asynchronous request to get the end-offsets for the partition
+     * with the epoch less than or equal to the epoch the partition last saw.
+     * <p/>
+     * Requests are grouped by Node for efficiency.
+     */
+    private List<NetworkClientDelegate.UnsentRequest> sendListOffsetsRequestsAndValidatePositions(
+            Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate) {
+
+        final Map<Node, Map<TopicPartition, SubscriptionState.FetchPosition>> regrouped =
+                offsetFetcherUtils.regroupFetchPositionsByLeader(partitionsToValidate);
+
+        long nextResetTimeMs = time.milliseconds() + requestTimeoutMs;
+        final List<NetworkClientDelegate.UnsentRequest> unsentRequests = new ArrayList<>();
+        regrouped.forEach((node, fetchPositions) -> {
+
+            if (node.isEmpty()) {
+                metadata.requestUpdate();
+                return;
+            }
+
+            NodeApiVersions nodeApiVersions = apiVersions.get(node.idString());
+            if (nodeApiVersions == null) {
+                // client.tryConnect(node);
```

Comment:
do we still want to keep this?

Reply:
i feel this should be part of the network client delegate

Reply:
Agree that the connection-related logic for reconnecting when needed should be in the poll. Will remove the comment . Just to double check, you were referring to the client.tryConnect only right? I'm leaving the logic for the early return because we do need to have access to the node api version in order to validate the protocol version the broker supports (and skip validation or not).

Reply:
Exactly, I was referring to tryConnect.  

When examine `handleInitiateApiVersionRequests -> doSend`. it seems like hasUsableOffsetForLeaderEpochVersion also checks for null, so this is a bit of double checking.  Is there anyway we could eliminate this if else? Basically, I think the reason old code invoke tryConnect is because it doesn't automatically polls the network client, which will attempt to update the apiVersion. See `handleInitiateApiVersionRequests -> doSend`


This is what i'm referring to.
```
    static boolean hasUsableOffsetForLeaderEpochVersion(NodeApiVersions nodeApiVersions) {
        ApiVersion apiVersion = nodeApiVersions.apiVersion(ApiKeys.OFFSET_FOR_LEADER_EPOCH);
        if (apiVersion == null)
            return false;

        return OffsetsForLeaderEpochRequest.supportsTopicPermission(apiVersion.maxVersion());
    }
    ```
    
    Sorry for being annoying...

Reply:
Agree that we can remove the tryConnect (done). But not quite getting the idea with removing the checks: here on the OffsetRequestManager we check 2 things:
1-  `nodeApiVersions != null`
2-  `nodeApiVersions.apiVersion(ApiKeys.OFFSET_FOR_LEADER_EPOCH) != null`. This is the ultimate goal here, needed to determine if we should skip validation, and the reason for check 1.

The doSend indeed checks the same 1-`nodeApiVersions != null, because it needs it to get `nodeApiVersions.latestUsableVersion()`. So I see that assuming that we do want the OffsetFetcherManager to keep doing the apiVersion check to skip offsets validation, both places should check for the nodeApiVersions for nulls. WDYT?

Reply:
sounds good. let's just remove the comment.